drivers/eeprom/eeprom_simulator.c

```diff
@@ -0,0 +1,309 @@
+
+	if ((offset + len) < config->size) {
+		return 1;
+	}
+
+	return 0;
+}
+
+static int eeprom_sim_read(struct device *dev, const off_t offset, void *data,
+			  const size_t len)
+{
+	if (!len) {
+		return 0;
+	}
+
+	if (!eeprom_range_is_valid(dev, offset, len)) {
+		LOG_WRN("attempt to read past device boundary");
+		return -EINVAL;
+	}
+
+	SYNC_LOCK();
+
+	STATS_INC(eeprom_sim_stats, eeprom_read_calls);
+	memcpy(data, EEPROM(offset), len);
+	STATS_INCN(eeprom_sim_stats, bytes_read, len);
+
+	SYNC_UNLOCK();
+
+#ifdef CONFIG_EEPROM_SIMULATOR_SIMULATE_TIMING
+	k_busy_wait(CONFIG_EEPROM_SIMULATOR_MIN_READ_TIME_US);
+	STATS_INCN(eeprom_sim_stats, eeprom_read_time_us,
+		   CONFIG_EEPROM_SIMULATOR_MIN_READ_TIME_US);
+#endif
+
+	return 0;
+}
+
+static int eeprom_sim_write(struct device *dev, const off_t offset,
+			   const void *data, const size_t len)
+{
+	const struct eeprom_sim_config *config = DEV_CONFIG(dev);
+
+	if (config->readonly) {
+		LOG_WRN("attempt to write to read-only device");
+		return -EACCES;
+	}
+
+	if (!len) {
+		return 0;
+	}
+
+	if (!eeprom_range_is_valid(dev, offset, len)) {
+		LOG_WRN("attempt to write past device boundary");
+		return -EINVAL;
+	}
+
+	SYNC_LOCK();
+
+	STATS_INC(eeprom_sim_stats, eeprom_write_calls);
+
+	bool data_part_ignored = false;
+
+	if (eeprom_sim_thresholds.max_write_calls != 0) {
+		if (eeprom_sim_stats.eeprom_write_calls >
+			eeprom_sim_thresholds.max_write_calls) {
+			return 0;
+		} else if (eeprom_sim_stats.eeprom_write_calls ==
+				eeprom_sim_thresholds.max_write_calls) {
+			if (eeprom_sim_thresholds.max_len == 0) {
+				return 0;
+			}
+
+			data_part_ignored = true;
+		}
+	}
+
+	for (u32_t i = 0; i < len; i++) {
+		if (data_part_ignored) {
+			if (i >= eeprom_sim_thresholds.max_len) {
+				return 0;
+			}
+		}
+
+		*(EEPROM(offset + i)) = *((u8_t *)data + i);
+	}
+
+	STATS_INCN(eeprom_sim_stats, bytes_written, len);
+
+#ifdef CONFIG_EEPROM_SIMULATOR_SIMULATE_TIMING
+	/* wait before returning */
+	k_busy_wait(CONFIG_EEPROM_SIMULATOR_MIN_WRITE_TIME_US);
+	STATS_INCN(eeprom_sim_stats, eeprom_write_time_us,
+		   CONFIG_EEPROM_SIMULATOR_MIN_WRITE_TIME_US);
+#endif
+
+	SYNC_UNLOCK();
+
+	return 0;
+}
+
+static size_t eeprom_sim_size(struct device *dev)
+{
+	const struct eeprom_sim_config *config = DEV_CONFIG(dev);
+
+	return config->size;
+}
+
+static const struct eeprom_driver_api eeprom_sim_api = {
+	.read = eeprom_sim_read,
+	.write = eeprom_sim_write,
+	.size = eeprom_sim_size,
+};
+
+static const struct eeprom_sim_config eeprom_sim_config_0 = {
+	.size = DT_INST_0_ZEPHYR_SIM_EEPROM_SIZE,
+	.readonly = DT_INST_0_ZEPHYR_SIM_EEPROM_READ_ONLY,
+};
+
+#ifdef CONFIG_ARCH_POSIX
+
+static int eeprom_mock_init(struct device *dev)
+{
+	if (eeprom_file_path == NULL) {
+		eeprom_file_path = default_eeprom_file_path;
+	}
+
+	eeprom_fd = open(eeprom_file_path, O_RDWR | O_CREAT, (mode_t)0600);
+	if (eeprom_fd == -1) {
+		posix_print_warning("Failed to open eeprom device file "
```

Comment:
Same comment as here: https://github.com/zephyrproject-rtos/zephyr/pull/20553/files/c631a4defa8ef73f9f130c0c2ab2c79b917b8556#diff-19518c7a115928db382df066c14dbe29

Reply:
@aescolar, do you mean use `posix_print_error()` or something else ?

Reply:
yep

Reply:
I am not sure if we should exit, just return an error that init failed.

Reply:
I guess it depends on what will happen later if the app does not handle it and tries to access the  eprom anyhow (segfault?). I have no strong opinion anyhow here.

Reply:
I would rather keep it as is, this allows you to check that the app handles it, which it should.